Content.Client/Popups/PopupSystem.cs

```diff
@@ -279,29 +307,32 @@ public abstract class PopupLabel
             public PopupType Type = PopupType.Small;
             public string Text { get; set; } = string.Empty;
             public float TotalTime { get; set; }
+            public int Repeats = 1;
         }
 
-        public sealed class CursorPopupLabel : PopupLabel
-        {
-            public ScreenCoordinates InitialPos;
-
-            public CursorPopupLabel(ScreenCoordinates screenCoords)
-            {
-                InitialPos = screenCoords;
-            }
-        }
-
-        public sealed class WorldPopupLabel : PopupLabel
+        public sealed class WorldPopupLabel(EntityCoordinates coordinates) : PopupLabel
         {
             /// <summary>
             /// The original EntityCoordinates of the label.
             /// </summary>
-            public EntityCoordinates InitialPos;
+            public EntityCoordinates InitialPos = coordinates;
+        }
 
-            public WorldPopupLabel(EntityCoordinates coordinates)
-            {
-                InitialPos = coordinates;
-            }
+        public sealed class CursorPopupLabel(ScreenCoordinates screenCoords) : PopupLabel
+        {
+            public ScreenCoordinates InitialPos = screenCoords;
         }
+
+        [UsedImplicitly]
+        private record struct WorldPopupData(
+            string Message,
+            PopupType Type,
+            EntityCoordinates Coordinates,
+            EntityUid? Entity);
+
+        [UsedImplicitly]
+        private record struct CursorPopupData(
+            string Message,
+            PopupType Type);
     }
```

Comment:
As these are being used as dictionary keys would be good to override IEquatable because otherwise it will use a slow-ass reflection to determine it.

Reply:
Record structs don't use reflection for equality, so this should be fine.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record#value-equality